File: leetcode/897. Increasing Order Search Tree/897.cpp
// https://helloacm.com/how-to-turn-a-binary-search-tree-into-a-increasing-order-search-tree/
// https://leetcode.com/problems/increasing-order-search-tree/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* increasingBST(TreeNode* root) {
        if (root == NULL) return NULL;
        stack<TreeNode*> st;
        TreeNode *p = root;
        TreeNode *ans = new TreeNode(-1);
        TreeNode *pre = ans;
        while (st.size() > 0 || p) {
            while (p) {
                st.push(p);
                p = p->left;
            }
            p = st.top();
            st.pop();
            p->left = NULL;
            pre->right = p;
            pre = p;
            p = p->right;
        }
        return ans->right;
    }
};

File: 897. Increasing Order Search Tree/897.py
# https://helloacm.com/teaching-kids-programming-inorder-traversal-algorithm-to-convert-binary-search-tree-to-increasing-order-search-tree/
# https://leetcode.com/problems/increasing-order-search-tree/
# EASY, RECURSION

class Solution:
    def increasingBST(self, root):
        def inorder(node):
            if not node:
                return
            yield from inorder(node.left)
            yield node.val
            yield from inorder(node.right)

        ans = cur = TreeNode(None)
        for v in inorder(root):
            cur.right = TreeNode(v)
            cur = cur.right
        return ans.right
